### PR TITLE
emscripten_toolchain_profiler, part 1/2

### PR DESCRIPTION
**Pull request by juj**
Implements an Emscripten toolchain profiler which can be used to profile the process-wide execution of the whole toolchain in action.

This work comes in two parts, the first is the profiler itself. The second part will add profiling blocks inside the toolchain, to populate the graphs that are printed with more fine grained information.

See here for an example of an incremental build of my 3D engine after touching one .cpp file: https://dl.dropboxusercontent.com/u/40949268/emcc/toolchain_profiler_logs/toolchain_profiler_example.html

The output looks a bit messy in the unzoomed view. To zoom in, grab with mouse from either the far left or right ends of the blue/purple background rectangle, and drag the selection area smaller.

For a second example that is a bit cleaner and which includes the code coming up in part 2, see

https://dl.dropboxusercontent.com/u/40949268/emcc/toolchain_profiler_logs/toolchain_profiler_example_2.html

especially the blue colored swimlane entries that highlight custom profiling blocks. (That graph also carries some custom mods to how Emscripten runs the multiprocessing pools)

Usage of the profiler is like this:

    // 1. Clean up files from old leftover runs, if any (not needed if no old log files are around
    emprofile.py --reset
    // 2. Run some commands with profiling enabled
    EM_PROFILE_TOOLCHAIN=1 emcc x.cpp -o x.o
    EM_PROFILE_TOOLCHAIN=1 emcc y.cpp -o y.js
    ....
    EM_PROFILE_TOOLCHAIN=1 emmake make
    // 3. Finish profiling by grabbing the logs from all commands run so far and graphing the results to a html file
    emprofile.py --graph --outfile=output_log_results.html

### CODE REVIEW

File: em++ [hunk 1]
```diff
@@ -5,12 +5,16 @@
...
+from tools.toolchain_profiler import *
+
 sys.argv += ['--emscripten-cxx']
 
 if sys.version_info.major == 2:
   import emcc
   if __name__ == '__main__':
+    ToolchainProfiler.record_process_start()
     emcc.run()
+    exit(0)
```
Comment (kripken) [inline em++]: why is this needed? it's implied anyhow

Comment (kripken) [inline em++]: oh, this isn't `sys.exit`, is it...

File: emscripten.py [hunk 2]
```diff
@@ -15,6 +15,7 @@
 from tools import jsrun, cache as cache_module, tempfiles
 from tools.response_file import read_response_file
 from tools.shared import WINDOWS
+from tools.toolchain_profiler import *
```
Comment (kripken) [inline emscripten.py]: why import if it's never used?

**Comment by kripken** [comment]
I have to say, I'm a little worried about

 * All the imports all around, adding unnecessary overhead when profiling isn't enabled.
 * The monkey-patching of methods, making the toolchain code more complex.

Optimally when profiling is not enabled, nothing would be monkeypatched, and the only overhead would be a single check. For example, at the top of `emcc.py` we could check if we are profiling, and if we are, then create that class and monkeypatch things. But otherwise, nothing would be done.

**Comment by kripken** [comment]
Or maybe that one check could be at the top of `tools/shared.py`.

**Comment by juj** [comment]
>Optimally when profiling is not enabled, nothing would be monkeypatched, and the only overhead would be a single check. For example, at the top of emcc.py we could check if we are profiling, and if we are, then create that class and monkeypatch things. But otherwise, nothing would be done.

That is how I initially had it, namely https://github.com/juj/emscripten/blob/toolchain_profiler/emcc.py#L36:

    if EM_PROFILE_TOOLCHAIN:
      Popen = ProfiledPopen
      check_call = profiled_check_call
      check_output = profiled_check_output
    else:
      Popen = subprocess.Popen
      check_call = subprocess.check_call
      check_output = subprocess.check_output

but I thought that having such a lengthy boilerplate in the beginning of each file would be a no-go. That is why I moved all of that inside `toolchain_profiler.py` so that each file would only need a single import line at the top.

Note that if profiling is not enabled, there is no monkeypatching, but `toolchain_profiler.py` provides the actual subprocess functions in https://github.com/kripken/emscripten/pull/4491/files#diff-5127fab9259fbcb028301023272e9085R217.

The second part will add constructs like this https://github.com/juj/emscripten/blob/toolchain_profiler/emcc.py#L1378, so it will be necessary to import the `toolchain_profiler.py` file even when not profiling, to get the dummy no-op object for `ToolchainProfiler.profile_block()`.

I'm not sure what would be the best way to modify this, but open for suggestions. What do you think?

File: emscripten.py (hunk 2 above)
Comment (juj) [inline emscripten.py]: It is used, but e.g. that file already accessed `Popen` without `subprocess.` prefix, here: https://github.com/juj/emscripten/blob/5b0abedb3ed83a294832575117a0ba06de6d9d21/emscripten.py#L142, so the import monkeypatches `Popen` to the profiled version.

Also in the second part of the PR, there are some profiling blocks that I added, so it gets more use, like https://github.com/juj/emscripten/blob/toolchain_profiler/emscripten.py#L156

File: em++ (hunk 1 above)
Comment (juj) [inline em++]: If not profiling, it will literally be `sys.exit`, as it comes from this line https://github.com/kripken/emscripten/pull/4491/files#diff-5127fab9259fbcb028301023272e9085R221. When profiling, it will be https://github.com/kripken/emscripten/pull/4491/files#diff-5127fab9259fbcb028301023272e9085R221.

**Comment by kripken** [comment]
What I am thinking is to do this in `tools/shared.py`:

````
if we-are-profiling:
  sys.exit = ..
  multiprocess.Popen = ...
````

In other words, monkeypatch **the system locations** when profiling. This avoids needing to change anything at all when not profiling, and all the code can keep using idiomatic `sys.exit` etc. normally.

It's a dynamic language so this kind of thing works ;)

**Comment by juj** [comment]
Oh I see. I'll give that a try tomorrow.

**Comment by juj** [comment]
Thanks, updated the code by overriding `sys` and `subprocess` directly.  How does it look now?

File: emprofile.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env python
```
Comment (kripken) [inline emprofile.py]: could `emprofile.py` be in `tools/`?

**Comment by kripken** [comment]
Overall looks good now, thanks.

Let's make sure all the new `import`s don't affect perf in the non-profiling case. Maybe build a large project (bullet or such) before after this PR, and measure that?

File: tools/toolchain_profiler.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
+import subprocess, os, time, sys, tempfile
+
+original_sys_exit = sys.exit
```
Comment (kripken) [inline tools/toolchain_profiler.py]: all this code up to the main `if` could go in the `if`, right? we don't need it when not profiling.

Comment (juj) [inline tools/toolchain_profiler.py]: check

File: emprofile.py (hunk 3 above)
Comment (juj) [inline emprofile.py]: check

**Comment by juj** [comment]
Built my 3D engine to test:

before:

    36.87
    36.29
    37.00
    avg 36.72 seconds

after

    37.13
    36.51
    37.02
    avg 36.89 seconds

so doesn't look like much of a difference.

**Comment by kripken** [comment]
Great, lgtm.
